### ISSUE TITLE
Abysmal scanning performance on large directories due to excessive scans for album artwork

### ISSUE THREAD

**Issue by Yetangitu** [airsonic-advanced/airsonic-advanced#828]
First, for those of you looking for a workaround for this problem: **Settings -> General -> Cover art source = External file only**

While testing the latest iteration of the cue-support change (#756) I noticed a large - and I mean large, as in "23 minutes to scan 18000 tracks, of which ~18 minutes is spent scanning a directory containing ~3200 files". I traced this down to calls to `findCoverart(children)` in `createMediaFile()`:

https://github.com/airsonic-advanced/airsonic-advanced/blob/b035842aa3a8aaad03f90513eefe555601548d7b/airsonic-main/src/main/java/org/airsonic/player/service/MediaFileService.java#L553

This gets called (`number of files in directory * some factor larger than 1`) times  _for each file in a directory_, exponentially increasing scanning time for larger directories. For reference I instrumented `JaudiotaggerParser` and counted the number of calls to `JaudiotaggerParser.isImageAvailable()` (which lies downstream from `findCoverArt()`) per file to visualise the problem. It shows that `isImageAvailable()` is called _5163 times for every file_ in a directory of 3178 files: 

```
awk '/isImageAvailable/ { print $7}' LOG|sort|uniq -c|sort -nr > sorted.LOG

   5163 isImageAvailable(/mnt/audio/music/Top_2000/2021/2021_12_31_23_59_38:Queen-Bohemian
   5163 isImageAvailable(/mnt/audio/music/Top_2000/2021/2021_12_31_23_53_46:NPO
   5163 isImageAvailable(/mnt/audio/music/Top_2000/2021/2021_12_31_23_49_01:NPO
   5163 isImageAvailable(/mnt/audio/music/Top_2000/2021/2021_12_31_23_47_42:Procol
   5163 isImageAvailable(/mnt/audio/music/Top_2000/2021/2021_12_31_23_39_10:Eagles-Hotel
   5163 isImageAvailable(/mnt/audio/music/Top_2000/2021/2021_12_31_23_27_08:Golden
...
...
...
```
This is done _for every file in this directory_:
```
$ grep 'Top_2000/2021' sorted.LOG |wc -l
3178
```
For this ~7000 file, ~18000 tracks collection there are _almost 18 million_ calls to `isImageAvailable()`:
```
$ grep isImageAvailable LOG|wc -l
17859279
```
Here's where the time goes during such a scan:

![image](https://user-images.githubusercontent.com/2754838/154924675-7905c2f6-fef6-40e5-ae85-f1c6012a7e3c.png)

Most time is spent inside `org.jaudiotagger.audio.mp3.MP3FILE` (and downstream calls), searching for artwork in `id3` tags.

The solution here seems obvious in that there should only be a single invocation of `findCoverArt(children)` per album, i.e. this call should be guarded by a conditional on whether this search has already been initiated during this scan,

Here's the call in context:
```java
                        // Look for cover art.
                        Path coverArt = findCoverArt(children);
                        if (coverArt != null) {
                            // placeholder to be persisted later
                            mediaFile.setArt(new CoverArt(-1, EntityType.MEDIA_FILE, folder.getPath().relativize(coverArt).toString(), folder.getId(), false));
                        }
```
BTW: what function does the "placeholder to be persisted later' fulfil? What is the purpose of the bare `-1` in this context? It seems to be a placeholder for the `entityId`, what does that mean in this context and why is this a bare number instead of a constant?

The main culprit behind these excessive calls seems to be `MediaScannerService:updateAlbum()`, specifically this line:

https://github.com/airsonic-advanced/airsonic-advanced/blob/b035842aa3a8aaad03f90513eefe555601548d7b/airsonic-main/src/main/java/org/airsonic/player/service/MediaScannerService.java#L419

Commenting out this line brings the scanning time back to reasonable proportions.

**Comment by randomnicode** [comment]
Thanks for profiling! This issue does seem bizarre, and perhaps someone may have overlooked something.

> The solution here seems obvious in that there should only be a single invocation of findCoverArt(children) per album, i.e. this call should be guarded by a conditional on whether this search has already been initiated during this scan,

There _should_ only be one invocation of findCoverArt(children) per album. Why do you think this is not the case?

Imagine a simple directory structure:
```
Root
  |- Artist
     |- Album
        |- Music1.mp3
        |- Music2.mp3
```

Now lets go through the call sequence:
- scanFile(Root)
  - createMediaFile -> skips all of the first branch of the isRegularFile conditional -> IS root, so skips most of the other branch, only the title is set
- scanFile(Artist)
  - Same as above, but isn't root, so the artist name is set
- scanFile(Album)
  - This runs the aforementioned check for findCoverArt(children)
- scanFile(Music1.mp3)
  - Won't run any check, uses the first branch of the isRegularFile conditional
- scanFile(Music2.mp3)
  - same as Music1.mp3

As you see, the invocation should happen only ONCE per album.

> BTW: what function does the "placeholder to be persisted later' fulfil? What is the purpose of the bare -1 in this context? It seems to be a placeholder for the entityId, what does that mean in this context and why is this a bare number instead of a constant?

These are two separate questions:
 - The -1 placeholder is just a random placeholder entityId, necessitated by the fact that for media files/artists/albums that are newly created, they don't even have an id during the scan (the id is assigned by the database upon persistence).
 - The placeholder means it is a place to park the coverart data structure temporarily during the scan even though it doesn't reside with the rest of the contents of that mediafile/artist/album object in the database. It is convenient because during scan we need to link these two together. Later on, we persist the rest of the object first, get assigned an id for it, and then persist the coverart data structure separately and remove it from the entity object.

Summarily:
- New file
- Scan it first -> create a mediafile object
- look for its cover art -> park that coverart within the mediafile object temporarily
- go do other stuff (like scan other files)
- at time of persistence: first persist the mediafile, and get assigned an id for the mediafile, then persist the coverart embedded within the mediafile using the new assigned id. then remove the coverart object

similar pattern for all 3 entities: media_file, artist, and album.

Do you have any other hypotheses for why the ridiculous number of calls to that function?

**Comment by Yetangitu** [comment]
What seems to happen is the following:

For a given directory `X`:

1. `mFS.getChildrenOf(X, ...)` is called, which leads to...
2. every file in `X` will be fed to `scanFile()`, which, since `musicFolder.getType() == MusicFolder.Type.MEDIA`, leads to...
3. `updateAlbum()` and `updateArtist()` being called, both of which do a check for artwork which, if the album/artist has no artwork, call `file.getParentOf()`:
https://github.com/airsonic-advanced/airsonic-advanced/blob/b035842aa3a8aaad03f90513eefe555601548d7b/airsonic-main/src/main/java/org/airsonic/player/service/MediaScannerService.java#L401
https://github.com/airsonic-advanced/airsonic-advanced/blob/b035842aa3a8aaad03f90513eefe555601548d7b/airsonic-main/src/main/java/org/airsonic/player/service/MediaScannerService.java#L454
4. ...both of which trickle down to `checkLastModified(X)`, which - given that this is called during a full scan, always calls...
5. `createMediaFile(X)`, which, because it is a directory, ends up running:
https://github.com/airsonic-advanced/airsonic-advanced/blob/6255670f31f1ccfde0a5a8d3b071339a81bf639c/airsonic-main/src/main/java/org/airsonic/player/service/MediaFileService.java#L553
6. ...which, if the thing is set to look for artwork in ID3-tags, will run `org.jaudiotagger.audio.mp3.MP3FILE` and its methods to try to extract artwork, which _takes forever_ in relative terms. Worse, these calls do not seem to actually return artwork, leading to those calls in `3` being repeated for every file in the directory.

Setting **Settings -> General -> Cover art source = External file only** works around this problem mostly due to to the fact that calls to `findFileCover()` take far less time than those to `findTagCover()` - it does not solve the problem of excessive calls to `findCoverArt()`.

I see a few possible solutions:
1.  mark the scan as being done after a single run, even if no artwork is found. It should not be repeated for the remaining invocations of `updateAlbum()` and `updateArtist()`
2. Instead of looking for artwork at every invocation of `updateAlbum()` and `updateArtist()` perhaps this search should be run separately, after all files are updated? Add an `updateArtwork()` function which looks for artwork for all albums and artists, run it before persisting albums. This is the solution I've used to fix the similar problem in the cue support change set (which I'll push one of these days).

**Comment by Yetangitu** [comment]
Here's a potential solution to this problem. There are two alternatives, one which restricts album art scans to "directory albums" by adding another `Map<String, Boolean> albumArtScanned` to the `scanFile()` signature, the other which simply uses `firstEncountered.get()` to restrict art scans. The first one is _fast_ but skips "albums" which are created while scanning directories containing tracks by various artists.

This also points at one of the backgrounds of this problem, the fact that the concept of "album" is handled differently depending on which part of the code you look at:

1. in `MediaFileService` an album is defined as a directory with children:
```java
            // Is this an album?
            if (!isRoot(mediaFile)) {
                try (Stream<Path> stream = Files.list(file)) {
                    List<Path> children = stream.parallel().collect(Collectors.toList());
                    Path firstChild = children.parallelStream()
                            .filter(x -> includeMediaFile(x))
                            .filter(x -> Files.isRegularFile(x))
                            .findFirst().orElse(null);

                    if (firstChild != null) {
                        mediaFile.setMediaType(MediaFile.MediaType.ALBUM);
```
2. in `MediaScannerService` an album is defined as a combination of work by an artist in a given directory:
```
        Album album = albums.compute(file.getAlbumName() + "|" + artist, (k,v) -> {
            Album a = v;

            if (a == null) {
                Album dbAlbum = albumDao.getAlbum(artist, file.getAlbumName());
                if (dbAlbum != null) {
                    a = albumsInDb.computeIfAbsent(dbAlbum.getId(), aid -> {
                        // reset stats when first retrieve from the db for new scan
                        dbAlbum.setDuration(0);
                        dbAlbum.setSongCount(0);
                        return dbAlbum;
                    });
                }
            }

            if (a == null) {
                a = new Album();
                a.setPath(file.getParentPath());
                a.setName(file.getAlbumName());
                a.setArtist(artist);
                a.setCreated(file.getChanged());
            }
```
This leads to the following sequence of events, for file `X` in directory `D` which contains a multitude of files by different artists.
  **a**. `MediaScannerService:scanFile()` is called for file `X`
  **b**. `updateAlbum()` is called with file `X` as parameter
  **c**. a search is made for an album based on `X.getAlbumName()`
  **d**. this call goes through `MetaDataParser.getMetaData()` which returns a guessed album name since the file does not contain metadata. The album name is set to the parent directory name, `D`
  **e**. A search is initiated for album art for an album named `D`
  **f**. this search triggers `createMediaFile(`D`) which, since this is a directory with children and thus "an album comprising all children as tracks", calls `findCoverArt(children)` as defined in **1**. This is wrong, the search for artwork should be limited to the files contained in the album created in **2**.
  ""g"". no artwork is found, `album.getArt()` will return `null` the next time around and the whole wild goose chase starts again.

To really solve this problem the naïve search for album art in `createMediaFile()` will have to be replaced by a content-aware search which _only_ searches files which actually belong to the album. Alternatively, and probably easier to implement is to create a specific `getAlbumArt()` method somewhere which is then called from `MediaFileScanner.updateAlbum()`.

**Comment by Yetangitu** [comment]
I'm currently using the first "solution" and have not noticed any problems with it.

Maybe the whole album-related logic should be removed from `createMediaFile()` and left to `MediaFileScanner.updateAlbum()`?

**Comment by randomnicode** [comment]
> 4. ...both of which trickle down to checkLastModified(X), which - given that this is called during a full scan, always calls...
> 5. createMediaFile(X), which, because it is a directory, ends up running:

No, fastcache (minimizeDiskAccess) is different from fullscan. This doesn't have to be called if doing a non full scan. in fact, it'll certainly never be called if fastcache is set to true, even during a scan

But your point is still very valid.

The artist/album treatment is weird and janky and arises from legacy concerns.
It is reflected in the REST calls and lucene indexes also (albumID3 is different from album)

However, there is a PR already out there to mitigate some of these concerns, and we can probably build upon it some of these excellent findings: https://github.com/airsonic-advanced/airsonic-advanced/pull/786

Can you check if its still hitting the same issues or how the PR may be modified within the new concepts and framework?

**Comment by Yetangitu** [comment]
#786 produces tons of `java.lang.IllegalArgumentException: DocValuesField "artist" appears more than once in this document (only one value is allowed per field)` errors so the comparison will not be totally relevant. That said, the slowdown seems to be similar to that in current master.

In what order are #831 / #756 and #786 to be merged? #786 does not apply cleanly on top of either so changes are needed. the opposite will be true as well (have not tried).

**Comment by randomnicode** [comment]
So, we think the issue is your use of full scan.

Every scan isn't usually a full scan. A full scan necessarily means rebuilding everything from scratch and thus scanning performance is expected to be slower. In this sequence of events you wrote about:

```
1. mFS.getChildrenOf(X, ...) is called, which leads to...
2. every file in X will be fed to scanFile(), which, since musicFolder.getType() == MusicFolder.Type.MEDIA, leads to...
3. updateAlbum() and updateArtist() being called, both of which do a check for artwork which, if the album/artist has no artwork, call file.getParentOf():
4. ..both of which trickle down to checkLastModified(X), which - given that this is called during a full scan, always calls...
5. createMediaFile(X), which, because it is a directory, ends up running:
6. ...which, if the thing is set to look for artwork in ID3-tags, will run org.jaudiotagger.audio.mp3.MP3FILE and its methods to try to extract artwork, which takes forever in relative terms. Worse, these calls do not seem to actually return artwork, leading to those calls in 3 being repeated for every file in the directory.
```

The critical piece here is the bottleneck at 4 and the conditional prior to 5. `createMediaFile(x)` in 5 won't be called unless you have a full scan in effect, which should be done rarely anyway.

However, even then, we can speed up the whole process tremendously by a simple change of passing in 'true' value for minimizeDiskAccess for checkLastModified, given that this change effectively (a) ensures that createMediaFile() never gets called and (b) causes even file modification timestamp checks to be bypassed so that there is only a DB lookup, and no disk read for the file.

Another speed up here would be the fact that cover art is being looked at again and again from the DB in the coverArtService, especially in scenarios where there is no cover art, because the cache doesn't cache nulls. For that we can just use an null-object pattern.

These changes are proposed in: https://github.com/airsonic-advanced/airsonic-advanced/pull/838
